Nodejs/Product/Nodejs/Project/MissingNodeInfoBar.cs

```diff
@@ -0,0 +1,107 @@
+using Microsoft.VisualStudio.Shell.Interop;
+using Microsoft.VisualStudioTools;
+
+namespace Microsoft.NodejsTools.Project
+{
+    internal sealed class MissingNodeInfoBar
+    {
+        private IServiceProvider serviceProvider;
+        private IVsInfoBarUIElement currentInfoBarElement;
+
+        public MissingNodeInfoBar(IServiceProvider serviceProvider)
+        {
+            this.serviceProvider = serviceProvider;
+        }
+
+        public void Show(NodejsProjectNode projectNode)
+        {
+            this.serviceProvider.GetUIThread().MustBeCalledFromUIThread();
+
+            var vsShell = (IVsShell)this.serviceProvider.GetService(typeof(SVsShell));
+            var infoBarUIFactory = (IVsInfoBarUIFactory)this.serviceProvider.GetService(typeof(SVsInfoBarUIFactory));
+
+            if (ErrorHandler.Failed(vsShell.GetProperty((int)__VSSPROPID7.VSSPROPID_MainWindowInfoBarHost, out var tmp)))
+            {
+                // we don't want to crash just because we can't show the error bar
+                return;
+            }
+            var infoBarHost = (IVsInfoBarHost)tmp;
+
+            // make sure we close the previous infobar
+            this.Close();
+
+            Action downloadNode = DownloadNode;
+            Action openProjectProps = ShowProjectProperties;
+
+            var actionItems = new[]
+            {
+                new InfoBarHyperlink(Resources.ConfigureProjectProperties, openProjectProps),
+                new InfoBarHyperlink(Resources.DownloadNodejs, downloadNode),
+            };
+
+            var infoBarModel = new InfoBarModel(Resources.NodejsNotInstalledInfoBar, actionItems, isCloseButtonVisible: true, image: KnownMonikers.StatusError);
+
+            uint eventCookie = 0;
+            this.currentInfoBarElement = infoBarUIFactory.CreateInfoBar(infoBarModel);
+            this.currentInfoBarElement.Advise(new InfoBarUIEvents(OnClose), out eventCookie);
+
+            infoBarHost.AddInfoBar(this.currentInfoBarElement);
+
+            void OnClose()
+            {
+                this.currentInfoBarElement.Unadvise(eventCookie);
+            }
+
+            void DownloadNode()
+            {
+                const string url = @"https://aka.ms/downloadnode";
+
+                VsShellUtilities.OpenBrowser(url, (uint)__VSOSPFLAGS.OSP_LaunchNewBrowser);
+            }
+
+            void ShowProjectProperties()
+            {
+                // open Project Properties
+                var logicalView = VSConstants.LOGVIEWID_Primary;
+                if (ErrorHandler.Succeeded(projectNode.GetGuidProperty(VSConstants.VSITEMID_ROOT, (int)VsHierarchyPropID.ProjectDesignerEditor, out var editorType)) &&
+                    ErrorHandler.Succeeded(projectNode.OpenItemWithSpecific(VSConstants.VSITEMID_ROOT, 0, ref editorType, "", ref logicalView, (IntPtr)(-1), out var frame)))
+                {
+                    frame?.Show();
+                }
+            }
+        }
+
+        public void Close()
```

Comment:
>public [](start = 8, length = 6)

private?

Reply:
Alternatively, consider inlining.

---
In reply to: [148164700](https://github.com/Microsoft/nodejstools/pull/1751#discussion_r148164700) [](ancestors = 148164700)

Reply:
Once it's inlined, there's no real reason to instantiate the type.  Show can just be a static method.

---
In reply to: [148165049](https://github.com/Microsoft/nodejstools/pull/1751#discussion_r148165049) [](ancestors = 148165049,148164700)